src/main/java/org/opensearch/security/filter/SecurityRestFilter.java

```diff
@@ -145,7 +150,54 @@ private boolean userIsSuperAdmin(User user, AdminDNs adminDNs) {
         return user != null && adminDNs.isAdmin(user);
     }
 
-    private boolean checkAndAuthenticateRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
+    private boolean authorizeRequest(RestHandler original, RestRequest request, RestChannel channel, User user) {
+
+        List<RestHandler.Route> restRoutes = original.routes();
+        Optional<RestHandler.Route> handler = restRoutes.stream()
+            .filter(rh -> rh.getMethod().equals(request.method()))
+            .filter(rh -> restPathMatches(request.path(), rh.getPath()))
+            .findFirst();
+        final boolean routeSupportsRestAuthorization = handler.isPresent() && handler.get() instanceof NamedRoute;
+        if (routeSupportsRestAuthorization) {
+            PrivilegesEvaluatorResponse pres = new PrivilegesEvaluatorResponse();
+            NamedRoute route = ((NamedRoute) handler.get());
+            // if actionNames are present evaluate those first
+            Set<String> actionNames = route.actionNames();
+            if (actionNames != null && !actionNames.isEmpty()) {
+                pres = evaluator.evaluate(user, actionNames);
+            }
+
+            // now if pres.allowed is still false check for the NamedRoute name as a permission
+            if (!pres.isAllowed()) {
+                String action = route.name();
+                pres = evaluator.evaluate(user, Set.of(action));
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug(pres.toString());
+            }
+            if (pres.isAllowed()) {
+                log.debug("Request has been granted");
+                auditLog.logGrantedPrivileges(user.getName(), request);
+            } else {
+                auditLog.logMissingPrivileges(route.name(), user.getName(), request);
+                String err;
+                if (!pres.getMissingSecurityRoles().isEmpty()) {
+                    err = String.format("No mapping for %s on roles %s", user, pres.getMissingSecurityRoles());
+                } else {
+                    err = String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user);
+                }
+                log.debug(err);
+                channel.sendResponse(new BytesRestResponse(RestStatus.UNAUTHORIZED, err));
+                return false;
+            }
+        }
+
+        // if handler is not an instance of NamedRoute then we pass through to eval at Transport Layer.
+        return true;
```

Comment:
If I am reading this correctly, we are always going to be evaluating both on the REST and trasnport layer correct? The only case where things would happen only in the REST layer would be if the request were unauthorized?

Reply:
> If I am reading this correctly, we are always going to be evaluating both on the REST and trasnport layer correct?

Yes.

> The only case where things would happen only in the REST layer would be if the request were unauthorized?

I'm not quite sure I understand this. Can you elaborate a little?

Reply:
I meant to clarify that the only time a request would end in the REST layer would be if it were unauthorized by the REST layer auth checker. Otherwise, it will always hit the transport layer checks.